Editable packaging solution

robobot3000 opened the pull request:
Adds #568 #569 #570 

Adds editable packing solution:
![oct-25-2016 13-05-13](https://cloud.githubusercontent.com/assets/800604/19681903/e7b19590-9ab3-11e6-8bc3-3dfdf2c21693.gif)

Left to do:
- [x] Use a dropdown to select a box type instead of a modal
- [x] Flat rate packages. Should we allow to select them?

CC @allendav @kellychoffman @DanReyLop

robobot3000 commented:
Spoke with @allendav - the flat rate packages are to be added in #642. This is ready for review.

kellychoffman commented:
This is what I see when I arrive at the Packages step:

![screen shot 2016-10-26 at 4 21 04 pm](https://cloud.githubusercontent.com/assets/5835847/19747541/8eea0a78-9b99-11e6-9ed5-518966b31a36.png)

I click 'add package' and there's only one option, `Envelope`:
![screen shot 2016-10-26 at 4 26 23 pm](https://cloud.githubusercontent.com/assets/5835847/19747588/d1b154f6-9b99-11e6-9a38-0c4bac0c3dc8.png)

Can we just show this by default?

One other issue here is the fact that the contents section is empty:

![screen shot 2016-10-26 at 4 29 00 pm](https://cloud.githubusercontent.com/assets/5835847/19747604/ebe06a92-9b99-11e6-8758-c957e753db6e.png)

I assume its because the item doesn't fit into this package. Could we display an error here? It allows me to move on to the last step and even purchase the label:

![screen shot 2016-10-26 at 4 29 00 pm](https://cloud.githubusercontent.com/assets/5835847/19747635/2b0931ae-9b9a-11e6-91c7-f0e34488a95e.png)

kellychoffman commented:
The Packages step doesn't auto expand once you have confirmed the destination address.

kellychoffman commented:
When I remove all packages, I see a floating `Saved for later`:

![screen shot 2016-10-26 at 4 59 48 pm](https://cloud.githubusercontent.com/assets/5835847/19748271/d9a87014-9b9d-11e6-9bbe-6f76e4c76f11.png)

robobot3000 commented:
You must have viewed an order with no packages or items in it. Not sure how that happened. Does the same order display any packages when you open it in a different branch?

> This is what I see when I arrive at the Packages step
> When I remove all packages, I see a floating Saved for later

As mentioned above, this shouldn't happen, but I added some messaging:
<img width="738" alt="screen shot 2016-10-27 at 14 23 10" src="https://cloud.githubusercontent.com/assets/800604/19765494/5c055852-9c51-11e6-87ac-23a6ef6cf191.png">
The floating `Saved for later` should be fixed also.

> One other issue here is the fact that the contents section is empty

I added a message for when the contents are empty:
<img width="699" alt="screen shot 2016-10-27 at 14 23 46" src="https://cloud.githubusercontent.com/assets/800604/19765565/c3859186-9c51-11e6-9f26-928010e082e9.png">

>  I click 'add package' and there's only one option, Envelope

The one option must be happening because you only have `Envelope` configured in you packaging manager.

> The Packages step doesn't auto expand once you have confirmed the destination address.

We don't autoexpand other sections when the ones before them were closed. The destination address wouldn't be expanded if it didn't need to be normalized. If we add autoexpand, then we should probably start at the origin address and then proceed from there?

jeffstieler commented:
This might be beyond the scope of this issues being resolved but.. it is possible to remove all of the package contents, and then be stuck.

We should have a way to add items to empty packages.

jeffstieler commented:
I'm having to scroll in almost all of the package modals.

See:
<img width="898" alt="screen shot 2016-10-27 at 12 13 32 pm" src="https://cloud.githubusercontent.com/assets/63922/19775778/826f48d4-9c3f-11e6-9ad0-1bc02d4f89c4.png">

jeffstieler commented on client/shipping-label/views/purchase/steps/packages/index.js:
```diff
@@ -1,56 +1,129 @@
+const PackagesStep = ( { openedPackageId, selected, all, unpacked, storeOptions, labelActions, errors, expanded, showItemMoveDialog, movedItemIndex, targetPackageId } ) => {
+	const packageIds = Object.keys( selected );
+	const itemsCount = packageIds.reduce( ( result, pId ) => ( result + selected[ pId ].items.length ), 0 );
+	const totalWeight = packageIds.reduce( ( result, pId ) => ( result + selected[ pId ].weight ), 0 );
```
Consider using lodash's [`sumBy()`](https://lodash.com/docs/4.16.4#sumBy) for this.

jeffstieler commented on client/shipping-label/views/purchase/steps/packages/index.js:
```diff
@@ -1,56 +1,129 @@
+	const packageIds = Object.keys( selected );
+	const itemsCount = packageIds.reduce( ( result, pId ) => ( result + selected[ pId ].items.length ), 0 );
+	const totalWeight = packageIds.reduce( ( result, pId ) => ( result + selected[ pId ].weight ), 0 );
+	const isValidWeight = packageIds.reduce( ( result, pId ) => ( result && 0 < selected[ pId ].weight ), true );
```
Can't this use the `totalWeight` value instead of a call to `.reduce()`?

jeffstieler commented on client/shipping-label/state/reducer.js:
```diff
@@ -140,7 +150,235 @@ reducers[ UPDATE_PACKAGE_WEIGHT ] = ( state, { packageId, value } ) => {
+	newPackages[ packageId ] = {
+		...oldPackage,
+		height, length, width,
+		weight: box.box_weight + oldPackage.items.reduce( ( sum, item ) => ( sum + item.weight ), 0 ),
```
Consider using lodash's [`sumBy()`](https://lodash.com/docs/4.16.4#sumBy) for this.

jeffstieler commented on client/shipping-label/state/reducer.js:
```diff
@@ -140,7 +150,235 @@ reducers[ UPDATE_PACKAGE_WEIGHT ] = ( state, { packageId, value } ) => {
+			},
+			rates: {
+				...state.form.rates,
+				values: _.mapValues( newPackages, ( pckg ) => ( pckg.service || '' ) ),
```
Same question about `rates.values` and `rates.available`.

jeffstieler commented on client/shipping-label/state/test/reducer.js:
```diff
@@ -0,0 +1,245 @@
+	afterEach( () => {
+		// make sure the state hasn't been mutated
+		// after each test
+		expect( initialState ).to.eql( {
```
Could using [`deepFreeze()`](https://github.com/substack/deep-freeze) on `initialState` be a way to avoid this?

robobot3000 replied:
The purpose of this is to test if the reducer methods don't mutate the state, so I actually want the `initialState` to not be frozen if this happens. But I removed the code duplication now, so this should be a bit easier to read.

jeffstieler commented on client/shipping-label/views/purchase/steps/packages/list.js:
```diff
@@ -1,78 +1,104 @@
+
+		if ( 'individual' === pckg.box_id ) {
+			individual.push(
+				<div className="wcc-packages-list__item" key={ pckgId }>
```
Let's keep this DRY - can you write a render method for package list items? (This is copied almost verbatim below for boxes/envelopes)

jeffstieler commented on client/shipping-label/views/purchase/steps/packages/list.js:
```diff
@@ -1,78 +1,104 @@
-			</div>
+	if ( individual.length ) {
+		individual.unshift(
+			<div className="wcc-packages-list__item wcc-packages-list__header" key="unpacked_header">{ __( 'Original packaging' ) }</div>
```
Not a huge deal, but another DRY candidate ("boxed" header above).

jeffstieler commented on client/shipping-label/views/purchase/steps/packages/move-item.js:
```diff
@@ -0,0 +1,101 @@
+					<p>{ __( 'Where would you like it moved?' ) }</p>
+					{ Object.values( _.mapValues( _.omitBy( selected, ( pckg, pckgId ) => ( pckgId === openedPackageId || 'individual' === pckg.box_id ) ), renderPackageOption ) ) }
+					{ ! openedPackage || 'individual' !== openedPackage.box_id ? renderRadioButton( 'individual', __( 'Ship in original packaging' ) ) : null }
+					{ '' !== openedPackageId ? renderRadioButton( '', __( 'Save for later' ) ) : null }
```
These three lines are really hard to read.

jeffstieler commented on client/shipping-label/views/purchase/steps/packages/package-info.js:
```diff
@@ -0,0 +1,90 @@
+			return ( <div className="wcc-package__desc">Package - { renderPackageDimensions( pckg, dimensionUnit ) }</div> );
+		}
+
+		const packageOptions = Object.values( _.mapValues( all, renderPackageOption ) );
```
You can use `_.map()` here instead of wrapping `_.mapValues()` with `Object.values()`.

robobot3000 commented on classes/class-wc-connect-shipping-label.php:
```diff
@@ -110,6 +110,23 @@ protected function get_packages( WC_Order $order ) {
+			$packages = $this->settings_store->get_packages();
+
+			if ( ! $packages ) {
+				return new stdClass();
```
The empty array will be serialized to json's `[]` which will lead to a React error, because it expects an object. Not sure if there's a way of working around that.

nabsul replied:
I wonder if we should just start using stdClass more consistently in our PHP code to avoid this issue altogether?

robobot3000 commented:
Thanks for reviewing @jeffstieler. I made the changes you suggested.
> I'm having to scroll in almost all of the package modals.
I tweaked the styling a bit, so the modal should be a bit taller, and its contents centered

> I just realized I can go to "saved for later" to find the items, but I found that by chance. There was nothing indicating to me that any items were "saved".
I think @kellychoffman is going to have a look at this once she gets her local branch to show packages and items.

kellychoffman commented:
> Package 1

In the Package manager, you can name your package. We don't seem to expose this anywhere. Instead of saying `Package 1` `Package 2` etc here, we should write out the name of the package so it matches what is in the dropdown and in the Package manager.

kellychoffman commented:
In the item contents and 'move' dialog, the product appears to be linked, but doesn't go anywhere. It should link to the edit product page.

kellychoffman commented:
If I add a new box with the intention of moving an item to that, I should be able to add it from the new box. This would be possible if we added a button next to the 'There are no items in this package.' message: 

<img width="715" alt="screen shot 2016-10-28 at 2 55 46 pm" src="https://cloud.githubusercontent.com/assets/5835847/19822082/eb97c39c-9d1e-11e6-995d-db412ba8ccb9.png">

kellychoffman commented:
> EDIT: I just realized I can go to "saved for later" to find the items, but I found that by chance. There was nothing indicating to me that any items were "saved".

Can we add a `count` to the right of the Saved packages link?

kellychoffman commented:
Question: why can't I change the box on a package with items, like I can on a package without items? This seems like the most convenient way to bulk move items from one package to another.

robobot3000 commented:
> In the Package manager, you can name your package. Instead of saying Package 1 Package 2 or Package - [dimensions here]
> Question: why can't I change the box on a package with items, like I can on a package without items? This seems like the most convenient way to bulk move items from one package to another.

Both questions are about the same thing: it seems the default packages selected by the customer were the flat rate packages. The labels UI displays `Package` when it can't match the preselected package to the saved shared packages. This could also happen when a custom package is selected, but later deleted. Maybe it should show `Unknown` instead?

> In the item contents listing and 'move' dialog, the product appears to be linked, but doesn't go anywhere. It should link to the edit product page.

It links to the product page for me. Looking at the code, this should always link to somewhere. What is the value of the `href` attribute on these links for you?

robobot3000 commented:
I also added the `Add Item` button.

kellychoffman commented:
> It links to the product page for me. Looking at the code, this should always link to somewhere. What is the value of the href attribute on these links for you?

Its just `<a target="_blank">` for me. Tried on a few different orders and product items and they are all empty.

kellychoffman commented:
> Its just <a target="_blank"> for me. Tried on a few different orders and product items and they are all empty.

I _do_ get links if I create a new product, purchase it, and then create an order.

jeffstieler commented:
Sorry to nitpick - but I think we should change the logic that calculates the "Box N" naming.

It should be calculated off packages only - right now it includes items in their original packaging. This is confusing if you start putting individually packages items into a box. You'll see `Box 4` for instance, and then as you move individual items to it, the number decrements.

![screen shot 2016-11-04 at 10 53 13 am](https://cloud.githubusercontent.com/assets/63922/20014601/416e20aa-a27d-11e6-83d7-5924f149c02b.png)

![screen shot 2016-11-04 at 10 53 32 am](https://cloud.githubusercontent.com/assets/63922/20014610/499011a8-a27d-11e6-81b0-c13880b0f9e8.png)

robobot3000 commented:
This is now fixed. I also tweaked the rates step to show the same package names as in the packages step
